effectful/handlers/llm/encoding.py

```diff
@@ -0,0 +1,261 @@
+import base64
+import dataclasses
+import io
+import numbers
+import typing
+from abc import ABC, abstractmethod
+from collections.abc import Callable
+
+import pydantic
+from litellm import ChatCompletionImageUrlObject
+from PIL import Image
+from pydantic import Field
+
+from effectful.ops.syntax import _CustomSingleDispatchCallable
+
+
+def _pil_image_to_base64_data(pil_image: Image.Image) -> str:
+    buf = io.BytesIO()
+    pil_image.save(buf, format="PNG")
+    return base64.b64encode(buf.getvalue()).decode("utf-8")
+
+
+def _pil_image_to_base64_data_uri(pil_image: Image.Image) -> str:
+    return f"data:image/png;base64,{_pil_image_to_base64_data(pil_image)}"
+
+
+class _Encodable[T, U](ABC):
+    t: type[U]
+
+    @classmethod
+    @abstractmethod
+    def encode(cls, t: T) -> U:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def decode(cls, t: U) -> T:
+        pass
+
+
+class Encodable[T](_Encodable[T, type]):
+    t = type
+
+
+@_CustomSingleDispatchCallable
+def type_to_encodable_type[T](
+    __dispatch: Callable[[type[T]], Callable[..., Encodable[T]]], ty: type[T]
+) -> Encodable[T]:
+    origin_ty = typing.get_origin(ty) or ty
+    return __dispatch(origin_ty)(ty)
+
+
+@type_to_encodable_type.register(object)
+def _type_encodable_type_object[T](ty: type[T]) -> Encodable[T]:
+    # Check if it's a dataclass and redirect to dataclass handler
+    if dataclasses.is_dataclass(ty):
+        return _type_encodable_type_dataclass(ty)
+    return _type_encodable_type_base(ty)
+
+
+@type_to_encodable_type.register(int)
+@type_to_encodable_type.register(float)
+@type_to_encodable_type.register(bool)
+@type_to_encodable_type.register(str)
+def _type_encodable_type_base[T](ty: type[T]) -> Encodable[T]:
+    class BaseEncodable(_Encodable[T, T]):
+        t = ty
+
+        @classmethod
+        def encode(cls, vl):
+            return vl
+
+        @classmethod
+        def decode(cls, vl: t) -> T:  # type: ignore
+            return vl
+
+    return typing.cast(Encodable[T], BaseEncodable())
+
+
+@type_to_encodable_type.register(numbers.Number)
+class EncodableNumber(_Encodable[numbers.Number, float]):
+    t = float
+
+    def __init__(self, _):
+        pass
+
+    @classmethod
+    def encode(cls, vl: numbers.Number) -> t:  # type: ignore
+        return float(vl)  # type: ignore
```

Comment:
I don't think this is right, not every `Number` can be invertibly coerced to a `float`. Maybe just remove the `numbers.Number` case for `type_to_encodable_type` for now, as long as base numeric types `int`, `float`, `bool`, and `complex` are all explicitly covered elsewhere.